angularSTEP/src/app/upload-recipe/upload-recipe.component.ts

```diff
@@ -76,41 +107,63 @@ export class UploadRecipeComponent {
+      if ((65 <= currentValue && currentValue <= 90) || (97 <= currentValue && currentValue <= 122)) {
+        return originalString.substring(i);
       }
+    }
+    if (startIndex == -1) {
       return '';
+    }
+    return '';
   }
 
+  formatArrays(originalArray: FormArray) {
+    for (let i = 0; i < originalArray.length; i++) {
+      originalArray.at(i).setValue(this.autoCapitalizeFirst(originalArray.at(i).value));
+    }
+  }
+
+  formatInstructions(originalArray: FormArray) {
+    for (let i = 0; i < originalArray.length; i++) {
+      originalArray.at(i).setValue(this.autoCapitalizeFirst(this.removeNumsAndSymbolsAtStart(originalArray.at(i).value)));
+    }
+  }
+
+  ngOnInit(): void {
+  }
+  
   onSubmit() {
-    if (this.autoCapitalizeFirst(this.recipeForm.value.extraInfo)) {
+    this.basicsFormGroup.controls.name.setValue(this.autoCapitalizeName(this.basicsFormGroup.value.name));
+    this.basicsFormGroup.controls.description.setValue(this.autoCapitalizeFirst(this.basicsFormGroup.value.description));
+    this.formatArrays(this.ingredientsArray);
+    this.formatArrays(this.toolsArray);
+    this.formatInstructions(this.instructionsArray);
+    this.extraFormGroup.controls.extraInfo.setValue(this.autoCapitalizeFirst(this.extraFormGroup.value.extraInfo));
+
+    if (this.extraFormGroup.value.extraInfo) {
       this.db.collection('recipes').add({
-        recipeName: this.autoCapitalizeName(this.recipeForm.value.name),
-        description: this.autoCapitalizeFirst(this.recipeForm.value.description),
-        ingredients: this.ingredients,
-        tools: this.tools,
-        instructions: this.instructions,
-        extraInfo: this.autoCapitalizeFirst(this.recipeForm.value.extraInfo),
-        difficulty: this.recipeForm.value.difficulty,
+        recipeName: this.basicsFormGroup.value.name,
+        difficulty: this.basicsFormGroup.value.difficulty,
+        description: this.basicsFormGroup.value.description,
+        ingredients: this.ingredientsArray.value,
+        tools: this.toolsArray.value,
+        instructions: this.instructionsArray.value,
         timestamp: Date.now(),
       });
     } else {
       this.db.collection('recipes').add({
-        recipeName: this.autoCapitalizeName(this.recipeForm.value.name),
-        description: this.autoCapitalizeFirst(this.recipeForm.value.description),
-        ingredients: this.ingredients,
-        tools: this.tools,
-        instructions: this.instructions,
-        difficulty: this.recipeForm.value.difficulty,
+        recipeName: this.basicsFormGroup.value.name,
+        difficulty: this.basicsFormGroup.value.difficulty,
+        description: this.basicsFormGroup.value.description,
+        ingredients: this.ingredientsArray.value,
+        tools: this.toolsArray.value,
+        instructions: this.instructionsArray.value,
+        extraInfo: this.extraFormGroup.value.extraInfo,
         timestamp: Date.now(),
       });
```

Comment:
Okay so here I think you accidentally switched when you add extraInfo. Right now it is adding extraInfo when there isn't any extraInfo and not adding it when you have extraInfo. I think a better model would be to add extraInfo no matter what. We want uniform documents and if some have extraInfo fields and others don't, it would be hard to know how and there could be possible accesses that aren't allowed. I would say that you could just have it so `extraInfo: this.extraFormGroup.value.extraInfo ? this.extraFormGroup.value.extraInfo : ''` That way you won't add null or something similar.